[HUDI-9468] Parquet Binary Copy at Rowgroup Level

zhangyue19921010 opened the pull request:
### Change Logs

details are as followed

### Impact

no

### Risk level (write none, low medium or high below)

low

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change. If not, put "none"._

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed


# Hudi Parquet Binary Stream Copy Optimization

## Background
- Hudi tables primarily use **Parquet format**.
- During **Clustering**, data undergoes redundant "column-to-row-to-column" transformations, involving:
  - Compression/decompression
  - Encoding/decoding
  - Row/column conversions
  - Format translations between compute engines and Hudi
- Similar inefficiencies exist in:
  - Low-update-ratio **compaction** scenarios
  - **CoW write amplification** with minimal data updates

## Optimization Approach
<img width="1314" alt="image" src="https://github.com/user-attachments/assets/aa3385d6-a90e-40b3-9a55-f45b7604c3d4" />

- **HoodieParquetRewriter**: A foundational implementation for:  
  ✅ **Column-chunk-level binary stream copying**  
  ✅ Directly copying unmodified Parquet column chunks to target files  
  ✅ Rewriting Parquet footer for fast replication  
- Applied to **Clustering scenarios** in this PR.

## Supported Capabilities
### 1. Schema Evolution
- Handles schema changes (e.g., adding fields, widening numeric types). **Behavior**:  
  - Uses the table's latest schema as output schema.  
  - Compares input file schemas against the latest schema.  
  - Fills missing fields with **null values** during binary copying.

### 2. Hudi Custom Metadata Merging
- Merges critical Hudi metadata from input files:  
  ```plaintext
  hoodie_min_record_key
  hoodie_max_record_key
  org.apache.hudi.bloomfilter
  hoodie_bloom_filter_type_code

- Process:
  - Collect metadata from all input files.
  - Compute aggregated values (e.g., global min/max keys and merge bloom filters).
  - Write merged metadata to output files.
 
## Future Roadmap
Extend binary copy support to Merge-on-Read (MoR) Compaction and CoW Upsert operations.

## Optimization Results

Using identical configurations, Clustering was performed on 100GB of data with both the default strategy and the binary stream copy strategy. The binary stream copy strategy achieved:

- **93% reduction in execution time** (from 18 minutes to 1.2 minutes)

- **95% reduction in computational workload** (from 28.7 task-hours to 1.3 task-hours)

**Before**
<img width="1482" alt="image (1)" src="https://github.com/user-attachments/assets/ac7800c7-f033-43c0-a5e8-277cf9cd2a77" />
![image (2)](https://github.com/user-attachments/assets/51835cf7-4d9e-4f21-b5e5-2d7a5fc3cad8)


**After**

<img width="1638" alt="image (3)" src="https://github.com/user-attachments/assets/c78d9201-0ab9-449d-9d51-3654f3e49fc7" />

![image (4)](https://github.com/user-attachments/assets/b5fa09c0-7448-48fb-b075-3e3a8acb4973)

Commit:
hoodie rewrite handler and clustering using rewrite handler

zhangyue19921010 committed:
fix ut

zhangyue19921010 commented:
> Please block one review from either me or @yihua .. to ensure the changes play well with how we are thinking about changing the spark writer path

Okay.

zhangyue19921010 committed:
fix ut

zhangyue19921010 committed:
fix ut

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/bloom/HoodieDynamicBoundedBloomFilter.java:
```diff
@@ -125,5 +126,42 @@ private void extractAndSetInternalBloomFilter(DataInputStream dis) throws IOExce
+    int needAddRowNum = targetMatrixLength - initMatrixLength;
+    if (needAddRowNum > 0) {
+      for (int i = 0; i < needAddRowNum; i++) {
+        this.internalDynamicBloomFilter.addRow();
```
Not sure if we can rebuild the filter in one shot instead of invoking the `addRow` multiple times?

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/io/storage/rewrite/HoodieFileRewriter.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.io.IOException;
+
+public interface HoodieFileRewriter {
```
we can just extend the existing `HoodieFileWriter` and add a new interface method: `HoodieFileWriter#deepCopy` and only add a impl for parquet.

vinothchandar replied:
Lets not overload the `FileWriter` with this feature, since its not clear if we can handle this for  all file types? FileWriter works at the record iterator level . IMO deepCopy/fastCopy like implementations work one level lower than record level iterators writers do.. 

I question if we need an interface esp in its current form where its just two methods that don't take any args. What other implementations do we want to have for the `HoodieFileRewriter`

vinothchandar replied:
The interface shoud at-least cover `rewrite(inputFiles, outputFiles, cfg)` type methods?

zhangyue19921010 replied:
Maybe this interface is useful here. Since parquet and orc both can implement Binary Copy
1. Binary copy parquet at row group level 
2. Binary copy orc at stripe level.

Also as VC mentioned. Change the function to 
` long binaryCopy(List<StoragePath> inputFilePaths, List<StoragePath> outputFilePath, MessageType writeSchema, Properties props) throws IOException;
`

TheR1sing3un commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkStreamCopyClusteringExecutionStrategy.java:
```diff
@@ -0,0 +1,148 @@
+ *  2. Sort function is not supported yet.
+ *  3. Each clustering group only has one task to write.
+ */
+public class SparkStreamCopyClusteringExecutionStrategy<T> extends SparkSortAndSizeExecutionStrategy<T> {
```
How about inherited from ` SingleSparkJobExecutionStrategy `? It seems that you inherited from `MultipleSparkJobExecutionStrategy` and rewrote `performClustering` to submit only one spark job, and then mapped it into individual tasks according to the `ClusteringGroupInfo`. This logic has actually been implemented in `SingleSparkJobExecutionStrategy`, and you only need to rewrite the execution logic of each group, `SingleSparkJobExecutionStrategy#performClusteringForGroup`

vinothchandar replied:
Good point

zhangyue19921010 replied:
changed.

danny0405 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkStreamCopyClusteringExecutionStrategy.java:
```diff
@@ -0,0 +1,148 @@
+ * PAY ATTENTION!!!
+ * IN THIS STRATEGY
+ *  1. Only support clustering for cow table.
+ *  2. Sort function is not supported yet.
```
+1, the naming is confusing.

vinothchandar replied:
+1 on a new strategy. if sorting is not supported, this should not extend the SortAndSize..Strategy.. Let's cleanup code in a separate PR if you prefer to get this landed in a good place.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/SingleFileHandleRewriteFactory.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+
+/**
+ * Rewrite all inputFilePaths related files into one huge file.
```
"rewrite" is lil confusing, since no data is changing.. 

"Combine all input file paths quickly into a single large file" ?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkStreamCopyClusteringExecutionStrategy.java:
```diff
@@ -0,0 +1,148 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkStreamCopyClusteringExecutionStrategy.class);
+
+  public SparkStreamCopyClusteringExecutionStrategy(
```
lets please add UTs for this class. and all the new files introduced

zhangyue19921010 replied:
added `hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/procedure/TestClusteringBinaryCopyStrategy.scala`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/bloom/HoodieDynamicBoundedBloomFilter.java:
```diff
@@ -125,5 +126,42 @@ private void extractAndSetInternalBloomFilter(DataInputStream dis) throws IOExce
   }
+
+  @Override
+  public void or(BloomFilter other) {
```
unit tests

zhangyue19921010 replied:
In TestHoodieFileMetadataMerger#testBloomFilter

vinothchandar replied:
Same. please move them to the UT for this class. I know it's a bit scattered now . but change starts somewhere :)

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/bloom/BloomFilter.java:
```diff
@@ -54,4 +54,10 @@ public interface BloomFilter {
+   * Performs a logical OR operations with other BloomFilter.
+   * @param other
+   */
+  void or(BloomFilter other);
```
unit tests.

zhangyue19921010 replied:
In TestHoodieFileMetadataMerger#testBloomFilter

vinothchandar replied:
@zhangyue19921010 can we move them to `TestBloomFilter` or the direct UT for the class.

zhangyue19921010 replied:
Add `TestBloomFilter#testBloomORMethod` cover `SIMPLE` and `DYNAMIC_V0` BloomFilterTypeCode

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/TypedProperties.java:
```diff
@@ -39,7 +39,7 @@ public TypedProperties() {
   }
 
-  protected TypedProperties(Properties defaults) {
+  public TypedProperties(Properties defaults) {
```
+1

zhangyue19921010 replied:
removed.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/io/storage/rewrite/HoodieFileMetadataMerger.java:
```diff
@@ -0,0 +1,84 @@
+ * limitations under the License.
+ */
+
+package org.apache.hudi.io.storage.rewrite;
```
lets keep this one level higher.. `hudi.io.rewrite` ? naming wise : lets stick to one concept across the PR. 

fastcopy or rewrite. and rename this `..hudi.io.fastcopy` ?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/io/storage/rewrite/HoodieFileMetadataMerger.java:
```diff
@@ -0,0 +1,84 @@
+
+  }
+
+  public Map<String, String> mergeMetaData(Map<String, String> metaMap) {
```
UT

zhangyue19921010 replied:
in `hudi-common/src/test/java/org/apache/hudi/io/storage/TestHoodieFileMetadataMerger.java`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/io/storage/rewrite/HoodieFileRewriterFactory.java:
```diff
@@ -0,0 +1,72 @@
+      HoodieConfig config,
+      HoodieFileMetadataMerger metadataMerger,
+      Schema writeSchemaWithMetaFields) throws IOException {
+    throw new UnsupportedOperationException();
```
msg?

zhangyue19921010 replied:
removed

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/io/storage/rewrite/HoodieFileRewriterFactory.java:
```diff
@@ -0,0 +1,72 @@
+    return factory.newFileRewriter(inputFilePaths, targetFilePath, conf, config, metadataMerger, writeSchemaWithMetaFields);
+  }
+
+  protected <T> HoodieFileRewriter newFileRewriter(
```
can we avoid this

zhangyue19921010 replied:
removed

vinothchandar commented on hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetFileRewriter.java:
```diff
@@ -0,0 +1,131 @@
+import java.util.Queue;
+import java.util.Set;
+
+public class HoodieParquetFileRewriter extends HoodieParquetRewriterBase implements HoodieFileRewriter {
```
is this adapted from parquet-mr/parquet-java?

zhangyue19921010 replied:
Yes, it is.

vinothchandar commented on hudi-hadoop-common/src/test/java/org/apache/hudi/parquet/io/TestHoodieParquetFileRewriter.java:
```diff
@@ -0,0 +1,459 @@
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+public class TestHoodieParquetFileRewriter {
```
can we add more tests from the hudi context.. covering meta fields, stats and so forth. bloom filters..

vinothchandar commented on hudi-hadoop-common/src/test/java/org/apache/hudi/parquet/io/TestHoodieParquetFileRewriter.java:
```diff
@@ -0,0 +1,459 @@
+    verifyCodec(outputFile, CompressionCodecName.GZIP);
+
+    // Verify the data are not changed
+    validateColumnData();
```
would n't some meta fields like `_hoodie_file_name` change.

vinothchandar commented on hudi-hadoop-common/src/test/java/org/apache/hudi/parquet/io/TestFileBuilder.java:
```diff
@@ -0,0 +1,188 @@
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+
+public class TestFileBuilder {
```
This seems like a test util and not an actual test? if this is borrowed from parquet.. can we move this along with other test utils.. and see if we can avoid copying over this much code.

vinothchandar commented on hudi-hadoop-common/src/test/java/org/apache/hudi/parquet/io/TestFile.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.apache.parquet.example.data.simple.SimpleGroup;
+
+public class TestFile {
```
same.. please consolidate this clas and TestFileBuilder into sth maintainable/minimize.

vinothchandar reviewed (commented):
Thanks for the valuable contribution! Mostly looked at the code structure and other things.. It'd be great if we can get these in order.. and add more tests..

vinothchandar reviewed (changes requested):
Will take a deeper pass into the `hudi.parquet.io` implementation after that.

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
Merge branch 'master' into stream-binary-copy

zhangyue19921010 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCreateHandle.java:
```diff
@@ -103,9 +103,9 @@ public HoodieCreateHandle(HoodieWriteConfig config, String instantTime, HoodieTa
-      this.fileWriter =
-          HoodieFileWriterFactory.getFileWriter(instantTime, path, hoodieTable.getStorage(), config,
-              writeSchemaWithMetaFields, this.taskContextSupplier, config.getRecordMerger().getRecordType());
+      this.fileWriter = initWriter
```
removed.

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

danny0405 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkBinaryCopyClusteringExecutionStrategy.java:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  private boolean supportBinaryStreamCopy() {
+    return this.getHoodieTable().getMetaClient().getTableType() == COPY_ON_WRITE;
```
For MOR with `INSERT` operation, it should also work?

gudladona replied:
Could you advise on what is needed to bring this capability to MOR with INSERT?

zhangyue19921010 replied:
@gudladona Sorry for missing this comment.
For Mor + insert, maybe we need to check all the clustering operations are base file only, so that we can trigger this binary copy. Will support in followed pr.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieFileBinaryCopier.java:
```diff
@@ -0,0 +1,40 @@
+ * 1. Merge multi-parquet files at rowgroup level
+ * 2. Merge multi-orc files at stripe level
+ */
+public interface HoodieFileBinaryCopier {
```
HoodieFileBinaryCopier -> HoodieFileBinaryCopyWriter?

danny0405 commented on hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetFileBinaryCopier.java:
```diff
@@ -0,0 +1,156 @@
+
+  // Routines to get reader of next input file and set up relevant states
+  private void initNextReader() {
+    if (reader != null) {
```
did we close the last reader properly?

danny0405 commented on pom.xml:
```diff
@@ -2604,6 +2609,7 @@
         <orc.flink.version>1.5.6</orc.flink.version>
         <flink.avro.version>1.11.4</flink.avro.version>
         <flink.format.parquet.version>1.12.2</flink.format.parquet.version>
+        <parquet.version>${flink.format.parquet.version}</parquet.version>
```
not sure why we need this, the `parquet.version` will finally be overridden in https://github.com/apache/hudi/blob/2db999a1a253fb222f1e4a397e46f92f2b1a846d/packaging/hudi-flink-bundle/pom.xml#L39

zhangyue19921010 replied:
it's about changed the parquet version for other model like hudi-hadoop-common. Without this config compile will failed with followed errors
```
....
  Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project hudi-hadoop-common: Compilation failure: Compilation failure: 
Error:  /home/runner/work/hudi/hudi/hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetBinaryCopyBase.java:[37,52] package org.apache.parquet.column.values.bloomfilter does not exist
Error:  /home/runner/work/hudi/hudi/hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetBinaryCopyBase.java:[39,33] package org.apache.parquet.crypto does not exist
Error:  /home/runner/work/hudi/hudi/hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetBinaryCopyBase.java:[40,33] cannot find symbol
Error:    symbol:   class BlockCipher
Error:    location: package org.apache.parquet.format
Error:  /home/runner/work/hudi/hudi/hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetBinaryCopyBase.java:[47,33] org.apache.parquet.hadoop.ColumnChunkPageWriteStore is not public in org.apache.parquet.hadoop; cannot be accessed from outside package
Error:  /home/runner/work/hudi/hudi/hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/HoodieParquetBinaryCopyBase.java:[54,38] cannot find symbol
Error:    symbol:   class CompressionConverter
Error:    location: package org.apache.parquet.hadoop.util
...
```

zhangyue19921010 replied:
Make sense Danny. Resolve this issue through https://github.com/apache/hudi/pull/13396

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
add more uts

zhangyue19921010 committed:
add more uts

zhangyue19921010 committed:
add more uts

zhangyue19921010 committed:
add more uts

Commit:
Merge branch 'master' into stream-binary-copy

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
Fix checkstyle

zhangyue19921010 committed:
Fix checkstyle

vinothchandar commented:
@zhangyue19921010 please mention me again, once you have addressed the feedback.

zhangyue19921010 commented:
> @zhangyue19921010 please mention me again, once you have addressed the feedback.

hi @vinothchandar all comments are addressed, added uts passed PTAL

danny0405 commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/HoodieDatasetBulkInsertHelper.scala:
```diff
@@ -150,14 +149,8 @@ object HoodieDatasetBulkInsertHelper
-        dataset.schema
-    }
+                 shouldPreserveHoodieMetadata: Boolean): HoodieData[WriteStatus] = {
+    val schema = AvroConversionUtils.alignFields(dataset.schema, new Schema.Parser().parse(writeConfig.getSchema))
```
Is this a enforcement for nexted data types?

zhangyue19921010 replied:
This data class correction is necessary. The existing "bulk insert as row" operation fetches the DataFrame's schema(may inconsistencies with table schema). This inconsistencies in the nullable properties of identical columns lead to file corruption/unreadability after binary copying. Therefore, we've implemented a nullable type correction for the bulk insert as row operation to resolve this issue.

danny0405 reviewed (approved):
+1, just some minor naming suggestions.

zhangyue19921010 committed:
take care of more conner cases and add uts

zhangyue19921010 committed:
take care of more conner cases and add uts

zhangyue19921010 commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/AvroConversionUtils.scala:
```diff
@@ -268,4 +267,106 @@ object AvroConversionUtils {
+  /**
+   * Recursively aligns the nullable property of hoodie table schema, supporting nested structures
+   */
+  def alignFields(sourceSchema: StructType, avroSchema: Schema): StructType = {
```
changed.

zhangyue19921010 committed:
take care of more conner cases and add uts

zhangyue19921010 commented:
> @zhangyue19921010 @danny0405 pls check the match of table_schema and parquet schema before do binary copy. Spark BulkInsert will produce parquets with required attribute； however those columns in hudi itself are optional。 Do binary copy will corrupt parquet file [https://github.com/apache/spark/blob/1a3ae66c6c48bb319f0798826085e694fa7d0b58/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389。](https://github.com/apache/spark/blob/1a3ae66c6c48bb319f0798826085e694fa7d0b58/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389%E3%80%82) Suggest: throw exception directly when found above mismatch.

Thanks for your reminding. We notice this so that we use `alignFieldsNullability` func to resolve nullable inconsistencies issue. However, for the existing data file, we still need to do some pre-verification, which I will add later.

zhangyue19921010 committed:
add more check before binary copy

danny0405 commented on hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestSparkBinaryCopyClusteringAndValidationMeta.java:
```diff
@@ -192,4 +219,115 @@ private void checkFileFooter(List<String> clusteredFiles, List<HoodieRecord> all
+    return Collections.singletonList(group);
+  }
+
+  static class DummyTaskContextSupplier extends TaskContextSupplier {
```
we have `LocalTaskContextSupplier` for testing.

zhangyue19921010 replied:
changed.

danny0405 commented on hudi-hadoop-common/src/main/java/org/apache/hudi/parquet/io/ParquetBinaryCopyChecker.java:
```diff
@@ -0,0 +1,195 @@
+      this.repetitions = repetitions;
+    }
+
+    public boolean isSchemaSupport() {
```
can we add some doc, what does this mean?

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 commented:
> @zhangyue19921010 @danny0405 pls check the match of table_schema and parquet schema before do binary copy. Spark BulkInsert will produce parquets with required attribute； however those columns in hudi itself are optional。 Do binary copy will corrupt parquet file [https://github.com/apache/spark/blob/1a3ae66c6c48bb319f0798826085e694fa7d0b58/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389。](https://github.com/apache/spark/blob/1a3ae66c6c48bb319f0798826085e694fa7d0b58/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java#L389%E3%80%82) Suggest: throw exception directly when found above mismatch.

Hi @xiarixiaoyao added all necessary checks in `supportBinaryStreamCopy`including 
1. Check if all input parquet files schema support binary copy    
    1) two level List structure check
    2)  Decimal types stored via INT32/INT64/INT96 check
2. Check if set of files contains only one type of BloomFilterTypeCode, including null
3. Check if the same column across these files has only one repetition type

zhangyue19921010 commented:
> @zhangyue19921010 nice work. Test with 1.7TB data, 300 columns binary copy achived `7.7 mins` vs normal write `35 mins`; and cpu cost cut down from `90% -> 60%`
> 
> i have add some buffer for inputStream
> 
> ```
> public class HoodieInputFile implements InputFile {
>    @Override
>    public SeekableInputStream newStream() throws IOException {
>        byte[] buffer = new byte[(int) stat.getLen()];
>        FSDataInputStream fsDataInputStream = fs.open(stat.getPath());
>       fsDataInputStream.readFully(buffer, 0, (int) stat.getLen);
>       return new DelegatingSeekableInputStream(new ByteBufferBackedInputStream(buffer)) {
>               public long getPos() ....
>               public void seek....
>         }
>   }
> }
> ```

Thanks @xiarixiaoyao for your help in verifying the functionality and performance of this PR. It's great to see the approximately **5x performance improvement**. cc: @vinothchandar @danny0405 ~

vinothchandar commented:
@zhangyue19921010 in future, please open PRs from your fork. and not a feature branch on hudi . :). just noticed.. 

Great to see the perf results. I will hop on the revised PR .

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/BinaryCopyHandleFactory.java:
```diff
@@ -0,0 +1,59 @@
+ * Combine all input file paths quickly into a single large file at block level.
+ * Without performing extra operations like data serialization/deserialization or compression/decompression.
+ */
+public class BinaryCopyHandleFactory<T, I, K, O>
```
like the name "binary copy" way better..  thanks!

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/BinaryCopyHandleFactory.java:
```diff
@@ -0,0 +1,59 @@
+      TaskContextSupplier taskContextSupplier) {
+    return new HoodieBinaryCopyHandle(
+        config,
+        commitTime,
```
nit. rename: commitTime -> instantTime

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/BinaryCopyHandleFactory.java:
```diff
@@ -0,0 +1,59 @@
+        commitTime,
+        partitionPath,
+        getNextFileId(fileIdPrefix),
+        hoodieTable,
```
rename: table

vinothchandar commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/AvroConversionUtils.scala:
```diff
@@ -268,4 +267,106 @@ object AvroConversionUtils {
+  /**
+   * Recursively aligns the nullable property of hoodie table schema, supporting nested structures
+   */
+  def alignFieldsNullability(sourceSchema: StructType, avroSchema: Schema): StructType = {
```
lets please ensure each method here has unit tests.

zhangyue19921010 replied:
Try my best to add UTs for core methods as much as possible to ensure code security :)
Add ut in TestAvroConversionUtils
```
  test("test alignFieldsNullability function") {
```

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/bloom/InternalDynamicBloomFilter.java:
```diff
@@ -210,13 +210,29 @@ public void readFields(DataInput in) throws IOException {
     matrix = tmp;
   }
 
+  protected void addRows(int size) {
```
UT

vinothchandar reviewed (changes requested):
Some cosmetic comments. I will review deeply as follow on.

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
code review

zhangyue19921010 commented:
@hudi-bot run azure

HuangZhenQiu commented:
Is the feature able to be used by Flink clustering? Any thing needs to be changed in hudi-flink-datasource?

vinothchandar reviewed (commented):
Getting closer! Please clarify few things and I can continue reviewing..

zhangyue19921010 commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/HoodieDatasetBulkInsertHelper.scala:
```diff
@@ -150,14 +149,8 @@ object HoodieDatasetBulkInsertHelper
-        dataset.schema
-    }
+                 shouldPreserveHoodieMetadata: Boolean): HoodieData[WriteStatus] = {
+    val schema = AvroConversionUtils.alignFieldsNullability(dataset.schema, new Schema.Parser().parse(writeConfig.getSchema))
```
### Why We Need `alignFieldsNullability`

Currently, Hudi retrieves the Table Schema from three sources in order of priority:  
1. Schema from commit file  
2. Schema from table config  
3. Schema from data file  

```java
private Option<Schema> getTableAvroSchemaInternal(boolean includeMetadataFields, Option<HoodieInstant> instantOpt) {
  Option<Schema> schema =
    (instantOpt.isPresent()
      ? getTableSchemaFromCommitMetadata(instantOpt.get(), includeMetadataFields)
      : getTableSchemaFromLatestCommitMetadata(includeMetadataFields))
      .or(() ->
        metaClient.getTableConfig().getTableCreateSchema()
          .map(tableSchema ->
            includeMetadataFields
              ? HoodieAvroUtils.addMetadataFields(tableSchema, hasOperationField.get())
              : tableSchema)
      )
      .or(() -> {
        Option<Schema> schemaFromDataFile = getTableAvroSchemaFromDataFileInternal();
        return includeMetadataFields
          ? schemaFromDataFile
          : schemaFromDataFile.map(HoodieAvroUtils::removeMetadataFields);
      });

  // TODO: Append partition columns in all read-paths
  if (metaClient.getTableConfig().shouldDropPartitionColumns() && schema.isPresent()) {
    return metaClient.getTableConfig().getPartitionFields()
      .map(partitionFields -> appendPartitionColumns(schema.get(), Option.ofNullable(partitionFields)))
      .or(() -> schema);
  }
  return schema;
}
```

However, when `using bulk insert as row,` the schema of the generated Parquet data files may have `inconsistent Optional/Required attributes` compared to the table schema. This occurs because bulk insert as row directly uses the DataFrame's schema as the write schema:

```scala
def bulkInsert(dataset: Dataset[Row],
               instantTime: String,
               table: HoodieTable[_, _, _, _],
               writeConfig: HoodieWriteConfig,
               arePartitionRecordsSorted: Boolean,
               shouldPreserveHoodieMetadata: Boolean,
               operation: WriteOperationType): HoodieData[WriteStatus] = {
  val schema = operation match {
    case WriteOperationType.CLUSTER =>
      alignNotNullFields(dataset.schema, new Schema.Parser().parse(writeConfig.getSchema))
    case _ =>
      dataset.schema // ← Uses raw DataFrame schema!
  }
```

**Generally**, we downplay the distinction between `Optional` and `Required` in schemas, making it seem like a minor issue. **However**, Binary Copy performs **pre-validation of schema alignment** – and these nullability mismatches cause Binary Copy to fail.(Without this validation, the copied files become unreadable due to deserialization failures.)

So that we introduced `alignFieldsNullability` to achieve two critical goals:  
1. ✅ Enable Binary Copy for **all Hudi write operations** (including `bulk insert`)  
2. ✅ Synchronize **Table Schema** and **Data File Schema** nullability  

---

### Experimental Verification  
The experiments below demonstrate how `bulk insert as row` creates data files with **inconsistent Optional/Required attributes** :

Spark bulk insert as row
```
Schema:
message spark_schema {
  optional binary _hoodie_commit_time (STRING);
  optional binary _hoodie_commit_seqno (STRING);
  optional binary _hoodie_record_key (STRING);
  optional binary _hoodie_partition_path (STRING);
  optional binary _hoodie_file_name (STRING);
  required int32 id;
  required int32 id2;
  required group maptype (MAP) {
    repeated group key_value {
      required binary key (STRING);
      optional binary value (STRING);
    }
  }
  required group structtype {
    optional binary name (STRING);
    optional int32 age;
  }
  required binary arraytype (STRING);
  required group newarraytype (LIST) {
    repeated group list {
      optional int32 element;
    }
  }
}
```

Spark insert/upsert
```
Schema:
message hoodie.app_jdr_normal_struct_and_map_a_d_d.app_jdr_normal_struct_and_map_a_d_d_record {
  optional binary _hoodie_commit_time (STRING);
  optional binary _hoodie_commit_seqno (STRING);
  optional binary _hoodie_record_key (STRING);
  optional binary _hoodie_partition_path (STRING);
  optional binary _hoodie_file_name (STRING);
  required int32 id;
  optional int32 id2;
  optional group maptype (MAP) {
    repeated group key_value (MAP_KEY_VALUE) {
      required binary key (STRING);
      optional binary value (STRING);
    }
  }
  optional group structtype {
    optional binary name (STRING);
    optional int32 age;
  }
  optional binary arraytype (STRING);
  optional group newarraytype (LIST) {
    repeated int32 array;
  }
}
```

Flink insert/upsert
```
Schema:
message flink_schema {
  optional binary _hoodie_commit_time (STRING);
  optional binary _hoodie_commit_seqno (STRING);
  optional binary _hoodie_record_key (STRING);
  optional binary _hoodie_partition_path (STRING);
  optional binary _hoodie_file_name (STRING);
  required int32 id;
  optional int32 id2;
  optional group maptype (MAP) {
    repeated group key_value {
      required binary key (STRING);
      optional binary value (STRING);
    }
  }
  optional group structtype {
    optional binary name (STRING);
    optional int32 age;
  }
  optional binary arraytype (STRING);
  optional group newarraytype (LIST) {
    repeated group list {
      optional int32 element;
    }
  }
}
```
**Overall**, Spark `insert`/`upsert` and Flink `insert`/`upsert` produce **consistent schemas**, while **Spark `bulk insert as row`** (and similar operations like Clustering) exhibit **inconsistent Optional/Required field attributes**.

zhangyue19921010 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkBinaryCopyClusteringExecutionStrategy.java:
```diff
@@ -0,0 +1,188 @@
+        .map(op -> {
+          String filePath = op.getDataFilePath();
+          if (!filePath.endsWith(PARQUET.getFileExtension())) {
+            return new ParquetBinaryCopyChecker.ParquetFileInfo(false, null, null);
```
Yes we need to fall back for not parquet format for now.

zhangyue19921010 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkConsistentBucketClusteringExecutionStrategy.java:
```diff
@@ -81,7 +80,7 @@ public HoodieData<WriteStatus> performClusteringWithRecordsAsRow(Dataset<Row> in
     Dataset<Row> repartitionedRecords = partitioner.repartitionRecords(inputRecords, numOutputGroups);
 
     return HoodieDatasetBulkInsertHelper.bulkInsert(repartitionedRecords, instantTime, getHoodieTable(), newConfig,
-        partitioner.arePartitionRecordsSorted(), shouldPreserveHoodieMetadata, WriteOperationType.CLUSTER);
```
No impact actually.

zhangyue19921010 commented on hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/client/functional/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -1017,6 +1017,14 @@ public void testSimpleClustering(boolean populateMetaFields) throws Exception {
+    String strategy = "org.apache.hudi.client.clustering.run.strategy.SparkBinaryCopyClusteringExecutionStrategy";
+    HoodieClusteringConfig config = createClusteringBuilder(true, 1).withClusteringExecutionStrategyClass(strategy).build();
+    testInsertAndClustering(config, true, true,
+        false, SqlQueryEqualityPreCommitValidator.class.getName(), COUNT_SQL_QUERY_FOR_VALIDATION, "");
```
Already have records check in `testInsertAndClustering` func when set complete clustering true

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
Merge branch 'master' into stream-binary-copy

zhangyue19921010 committed:
code review

zhangyue19921010 committed:
fix ut

zhangyue19921010 commented:
Comments addressed And all ci passed. PTAL @vinothchandar :)

vinothchandar commented:
@shangxinli @zhangyue19921010  : you both got the follow ups?
